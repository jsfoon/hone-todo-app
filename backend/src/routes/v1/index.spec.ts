import { describe, it, expect, afterAll } from "bun:test";

import todosApp from "./todo";

import { ISelectTodosQueryResult } from "../../models/todo/index.queries";

describe("todos", () => {
  // afterAll(async () => {
  //   const res = await todosApp.request("/");
  //   const result = (await res.json()) as { todos: ISelectTodosQueryResult[] };
  //   await Promise.all(
  //     result.todos.map(async (todo) => {
  //       await todosApp.request(`/${todo.id}`, {
  //         method: "DELETE",
  //       });
  //     })
  //   );
  // });
  it("GET /todos", async () => {
    const res = await todosApp.request("/");
    expect(res.status).toBe(200);
    const result = (await res.json()) as { todos: ISelectTodosQueryResult[] };
    expect(Array.isArray(result.todos)).toBe(true);
  });

  it("POST /todos", async () => {
    const res = await todosApp.request("/create", {
      method: "POST",
      body: JSON.stringify({ title: "create todo" }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    expect(res.status).toBe(200);
    const result = (await res.json()) as { todo: ISelectTodosQueryResult };
    expect(result.todo.title).toBe("create todo");
  });

  it("POST /todos with number", async () => {
    const res = await todosApp.request("/create", {
      method: "POST",
      body: JSON.stringify({ title: 123 }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    expect(res.status).toBe(400);
    const result = await res.json();
    expect(result).toMatchSnapshot();
  });
  it("PUT /todos/:id", async () => {
    const createRes = await todosApp.request("/create", {
      method: "POST",
      body: JSON.stringify({ title: "pre update todo" }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    expect(createRes.status).toBe(200);
    const createResult = (await createRes.json()) as {
      todo: ISelectTodosQueryResult;
    };
    expect(createResult.todo.title).toBe("pre update todo");
    const res = await todosApp.request(`/${createResult.todo.id}`, {
      method: "PUT",
      body: JSON.stringify({ title: "update todo" }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    expect(res.status).toBe(200);
    const result = (await res.json()) as { todo: ISelectTodosQueryResult };
    expect(result.todo.title).toBe("update todo");
  });

  it("PUT /todos/:id with string", async () => {
    const res = await todosApp.request(`/qwe`, {
      method: "PUT",
      body: JSON.stringify({ title: "update todo" }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    expect(res.status).toBe(400);
    const content = await res.text();
    expect(content).toMatchSnapshot();
  });

  it("DELETE /todos/:id", async () => {
    const createRes = await todosApp.request("/create", {
      method: "POST",
      body: JSON.stringify({ title: "pre delete todo" }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    expect(createRes.status).toBe(200);
    const createResult = (await createRes.json()) as {
      todo: ISelectTodosQueryResult;
    };

    const res = await todosApp.request(`/${createResult.todo.id}`, {
      method: "DELETE",
    });
    expect(res.status).toBe(200);
    const result = (await res.json()) as { todo: ISelectTodosQueryResult };
    expect(result.todo.title).toBe("pre delete todo");
    expect(result.todo.id).toBe(createResult.todo.id);
  });

  it("DELETE /todos/:id with string", async () => {
    const res = await todosApp.request(`/qwe`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
      },
    });
    expect(res.status).toBe(400);
    const content = await res.text();
    expect(content).toMatchSnapshot();
  });
});
