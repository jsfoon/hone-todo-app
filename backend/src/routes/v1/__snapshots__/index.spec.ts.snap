// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`todos PUT /todos/:id with string 1`] = `"{"success":false,"error":{"issues":[{"validation":"regex","code":"invalid_string","message":"Invalid","path":["id"]}],"name":"ZodError"}}"`;

exports[`todos POST /todos with number 1`] = `
{
  "error": {
    "issues": [
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Expected string, received number",
        "path": [
          "title",
        ],
        "received": "number",
      },
    ],
    "name": "ZodError",
  },
  "success": false,
}
`;

exports[`todos DELETE /todos/:id with string 1`] = `"{"success":false,"error":{"issues":[{"validation":"regex","code":"invalid_string","message":"Invalid","path":["id"]}],"name":"ZodError"}}"`;
